src/compiler/checker.ts

```diff
@@ -5351,17 +5355,40 @@ module ts {
                 }
                 // Target type is type of prototype property
                 let prototypeProperty = getPropertyOfType(rightType, "prototype");
-                if (!prototypeProperty) {
-                    return type;
+                if (prototypeProperty) {
+                    let targetType = getTypeOfSymbol(prototypeProperty);
+                    if (targetType !== anyType) {
+                        // Narrow to the target type if it's a subtype of the current type
+                        if (isTypeSubtypeOf(targetType, type)) {
+                            return targetType;
+                        }
+                        // If the current type is a union type, remove all constituents that aren't subtypes of the target.
+                        if (type.flags & TypeFlags.Union) {
+                            return getUnionTypeOfSubtypeConstituents(<UnionType>type, targetType);
+                        }
+                    }
                 }
-                let targetType = getTypeOfSymbol(prototypeProperty);
-                // Narrow to target type if it is a subtype of current type
-                if (isTypeSubtypeOf(targetType, type)) {
-                    return targetType;
+                // Target type is type of constructor signiture
+                let constructSignatures: Signature[];
+                if (rightType.flags & TypeFlags.Interface) {
+                    constructSignatures = (<InterfaceTypeWithDeclaredMembers>rightType).declaredConstructSignatures;

```

Comment:
getSignaturesOfType will also include signatures of the base type as well, which may pollute the resulting type. So actually, you do want to use declaredConstructSignatures. However, rather than casting, you need to ensure the signatures are computed, by accessing it like this:
```ts
resolveDeclaredMembers(rightType).declaredConstructSignatures
```